infra/build/functions/update_build_status.py

```diff
@@ -0,0 +1,118 @@
+def get_last_build(build_ids):
+  """Returns build object for the last finished build of project."""
+  credentials, image_project = google.auth.default()
+  cloudbuild = build('cloudbuild', 'v1', credentials=credentials)
+
+  for build_id in reversed(build_ids):
+    project_build = cloudbuild.projects().builds().get(projectId=image_project,
+                                                       id=build_id).execute()
+    if project_build['status'] == 'WORKING':
+      continue
+
+    if not builds_status.upload_log(build_id):
+      continue
+    return project_build
+
+  return None
+
+
+def update_build_status(build_tag_suffix, status_filename):
+  """Update build statuses."""
+  successes = []
+  failures = []
+  builds = BuildsHistory.query()
+  for project_build in builds:
+    if project_build.build_tag_suffix != build_tag_suffix:
+      continue
+    last_build = get_last_build(project_build.build_ids)
+    if not last_build:
+      logging.error('Failed to get last build for project %s', project_build.project)
+      continue
+
+    if last_build['status'] == 'SUCCESS':
+      successes.append({
+          'name': project_build.project,
+          'build_id': last_build['id'],
+          'finish_time': last_build['finishTime'],
+          'success': True,
+      })
+    else:
+      failures.append({
+          'name': project_build.project,
+          'build_id': last_build['id'],
+          'finish_time': last_build['finishTime'],
+          'success': False,
+      })
+
+  builds_status.upload_status(successes, failures, status_filename)
+
+
+# pylint: disable=no-member
+def update_status(event, context):
+  """Entry point for cloud function to update build statuses and badges."""
+  del event, context #unused
+
+  with ndb.Client().context():
+    update_build_status(build_project.FUZZING_BUILD_TAG,
+                        status_filename='status.json')
+    update_build_status(build_and_run_coverage.COVERAGE_BUILD_TAG,
+                        status_filename='status-coverage.json')
+
+    for project in Project.query():
+      build_history_query = BuildsHistory.query(
+          BuildsHistory.project == project.name,
+          BuildsHistory.build_tag_suffix == build_project.FUZZING_BUILD_TAG)
+
+      build_history = build_history_query.get()
+      if not build_history:
+        continue
+      last_build = get_last_build(build_history.build_ids)
+
+      if not last_build:
+        continue
+
+      coverage_build_history_query = BuildsHistory.query(
+          BuildsHistory.project == project.name,
+          BuildsHistory.build_tag_suffix == build_and_run_coverage.
+          COVERAGE_BUILD_TAG)
+
+      coverage_build_history = coverage_build_history_query.get()
+      if not coverage_build_history:
+        continue
+
+      last_coverage_build = get_last_build(coverage_build_history.build_ids)
+      if not last_coverage_build:
+        continue
+
+      builds_status.update_build_badges(project.name, last_build,
```

Comment:
Rather than do all the stuff in lines 91-115 again, we can improve things over the initial implementation by getting update_build_status to return the information we need. 

e.g. it can return a dict of project name -> build success, and can be used to make the builds_status.update_build_badges call.